File: python/2023/day01.py
def replace_nums(input: str) -> str:
    return input \
        .replace("one", "o1e") \
        .replace("two", "t2o") \
        .replace("three", "t3e") \
        .replace("four", "f4r") \
        .replace("five", "f5e") \
        .replace("six", "s6x") \
        .replace("seven", "s7n") \
        .replace("eight", "e8t") \
        .replace("nine", "n9e")


def solve(line: str) -> int:
    res = ""

    for ch in line:
        if ch.isdigit():
            res += ch
            break

    for ch in reversed(line):
        if ch.isdigit():
            res += ch
            break

    return int(res)


def part1(input: str) -> int:
    return sum(solve(line) for line in input.splitlines())


def part2(input: str) -> int:
    return sum(solve(line) for line in replace_nums(input).splitlines())


TEST_INPUT1 = """1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"""

TEST_INPUT2 = """two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen"""

PART1_TESTS = [
    (TEST_INPUT1, 142),
]

PART2_TESTS = [
    (TEST_INPUT2, 281),
]

File: rust/2023/src/bin/day01.rs
fn solve(line: &str) -> u32 {
    let numbers = line
        .chars()
        .filter_map(|c| c.to_digit(10))
        .collect::<Vec<_>>();

    numbers[0] * 10 + numbers[numbers.len() - 1]
}

fn replace_numbers(line: &str) -> String {
    line.replace("one", "o1e")
        .replace("two", "t2o")
        .replace("three", "t3e")
        .replace("four", "f4r")
        .replace("five", "f5e")
        .replace("six", "s6x")
        .replace("seven", "s7n")
        .replace("eight", "e8t")
        .replace("nine", "n9e")
}

struct Solution {
    lines: Vec<String>,
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        Ok(Self {
            lines: input.lines().map(String::from).collect(),
        })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self.lines.iter().fold(0, |acc, line| acc + solve(line)))
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self
            .lines
            .iter()
            .map(|line| replace_numbers(line))
            .fold(0, |acc, line| acc + solve(&line)))
    }
}

aoc::solution!();

#[cfg(test)]
mod test {
    use super::{Solution, Solver};

    const INPUT1: &str = r"1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet";

    const INPUT2: &str = r"two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen";

    #[test]
    fn test_part1() {
        let mut solution = Solution::new(INPUT1).unwrap();
        let answer = solution.part1().unwrap().to_string();
        assert_eq!(answer, "142");
    }

    #[test]
    fn test_part2() {
        let mut solution = Solution::new(INPUT2).unwrap();
        let answer = solution.part2().unwrap().to_string();
        assert_eq!(answer, "281");
    }
}
